File: 2_ReactWeb/src/services/network/networkSettings.ts
import { fetchJson } from "../http/httpClient";

export type NetworkConnectionMode = "system" | "direct" | "custom_proxy";
export type ProxyScheme = "http" | "https" | "socks5";
export type BackendPortMode = "auto" | "fixed";

export type NetworkSettings = {
  connection_mode: NetworkConnectionMode;
  proxy_scheme: ProxyScheme;
  proxy_host: string;
  proxy_port: number;
  connect_timeout_seconds: number;
  read_timeout_seconds: number;
  stream_timeout_seconds: number;
  backend_port_mode: BackendPortMode;
  fixed_backend_port: number;
};

export type NetworkSettingsResponse = {
  settings: NetworkSettings;
  default_settings: NetworkSettings;
  updated_at: string | null;
  backend_port_restart_required: boolean;
};

export type NetworkDiagnosticResponse = {
  ok: boolean;
  target: string;
  status_code: number | null;
  elapsed_ms: number;
  error: string | null;
};

export function getNetworkSettings() {
  return fetchJson<NetworkSettingsResponse>("/api/network/settings");
}

export function saveNetworkSettings(settings: NetworkSettings) {
  return fetchJson<NetworkSettingsResponse>("/api/network/settings", {
    body: JSON.stringify({ settings }),
    method: "PUT",
  });
}

export function diagnoseGithubConnection() {
  return fetchJson<NetworkDiagnosticResponse>("/api/network/diagnostics/github", {
    method: "POST",
  });
}

File: 1_PythonServer/app/domain/network_settings.py
from dataclasses import dataclass
from enum import StrEnum


class NetworkConnectionMode(StrEnum):
    SYSTEM = "system"
    DIRECT = "direct"
    CUSTOM_PROXY = "custom_proxy"


class ProxyScheme(StrEnum):
    HTTP = "http"
    HTTPS = "https"
    SOCKS5 = "socks5"


class BackendPortMode(StrEnum):
    AUTO = "auto"
    FIXED = "fixed"


@dataclass(frozen=True, slots=True)
class NetworkSettings:
    connection_mode: NetworkConnectionMode
    proxy_scheme: ProxyScheme
    proxy_host: str
    proxy_port: int
    connect_timeout_seconds: float
    read_timeout_seconds: float
    stream_timeout_seconds: float
    backend_port_mode: BackendPortMode
    fixed_backend_port: int
    updated_at: str | None = None


DEFAULT_NETWORK_SETTINGS = NetworkSettings(
    connection_mode=NetworkConnectionMode.SYSTEM,
    proxy_scheme=ProxyScheme.HTTP,
    proxy_host="127.0.0.1",
    proxy_port=7897,
    connect_timeout_seconds=10.0,
    read_timeout_seconds=120.0,
    stream_timeout_seconds=300.0,
    backend_port_mode=BackendPortMode.AUTO,
    fixed_backend_port=18000,
)
